src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -4,26 +4,615 @@
+   * @param fieldRelative Whether the provided x and y speeds are relative to the field.
+   */
+  public void keepHeading(double x, double y, boolean fieldRelative) {
+    double currentHeadingDegrees = getHeadingDegrees();
+    double headingDifferenceDegrees = currentHeadingDegrees - targetHeadingDegrees;
+    double offsetHeadingDegrees = MathUtil.inputModulus(headingDifferenceDegrees, -180, 180);
+
+    double pidRotation =
+        DriveCal.ROTATE_TO_TARGET_PID_CONTROLLER.calculate(offsetHeadingDegrees, 0.0);
+    double ffRotation = Math.signum(offsetHeadingDegrees) * DriveCal.ROTATE_TO_TARGET_FF;
+
+    double desiredRotation = pidRotation - ffRotation;
+
+    if (Math.abs(desiredRotation) < DriveCal.ROTATION_DEADBAND_THRESHOLD) {
+      desiredRotation = 0;
+    }
+
+    drive(x, y, desiredRotation, fieldRelative);
+  }
+
+  public int convertCardinalDirections(int povAngleDeg) {
+    // change d-pad values for left and right to specified angle
+    if (povAngleDeg == 270) {
+      povAngleDeg += 77;
+    } else if (povAngleDeg == 90) {
+      povAngleDeg -= 77;
+    }
+    // targetHeadingDegrees is counterclockwise so need to flip povAngle
+    povAngleDeg = 360 - povAngleDeg;
+    return povAngleDeg;
+  }
+
+  /**
+   * Determines whether to rotate according to input or to run the keep heading code, by checking if
+   * the (already deadbanded) rotation input is equal to 0.
+   *
+   * @param x Desired speed of the robot in the x direction (forward), [-1,1].
+   * @param y Desired speed of the robot in the y direction (sideways), [-1,1].
+   * @param rot Desired angular rate of the robot, [-1,1].
+   * @param fieldRelative Whether the provided x and y speeds are relative to the field.
+   * @param povAngleDeg Get the angle in degrees of the D-pad (clockwise, -1 means POV not pressed).
+   */
+  public void rotateOrKeepHeading(
+      double x, double y, double rot, boolean fieldRelative, int povAngleDeg) {
+    if (povAngleDeg != -1) {
+      targetHeadingDegrees = convertCardinalDirections(povAngleDeg);
+      keepHeading(x, y, fieldRelative);
+    } else if (rot == 0) {
+      keepHeading(x, y, fieldRelative);
+    } else {
+      targetHeadingDegrees = getHeadingDegrees();
+      drive(x, y, rot, fieldRelative);
+    }
+  }
+
+  public void setForward() {
+    frontLeft.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    frontRight.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    rearLeft.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    rearRight.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+  }
+
+  public WPI_Pigeon2 getGyro() {
+    return gyro;
+  }
+
+  /** Taken from Github */
+  public Command followTrajectoryCommand(PathPlannerTrajectory traj, boolean isFirstPath) {
+
+    return new SequentialCommandGroup(
+            new RunCommand(() -> setForward())
```

Comment:
We should reconsider this? Not every path goes forward/backward

Reply:
Is there a reasonable way to get the starting angle of the path? Or are you proposing we just remove that line without replacing it with anything?

Reply:
I think maybe just remove for now